### PR TITLE
Algorithmically better EConstr view function.

### PR DESCRIPTION
**Pull request by ppedrot**
Instead of substituting instance arguments eagerly, we use a delayed substitution to postpone the actual evaluation. The previous algorithm was quadratic in the length of evar chains in the worst case, it should now be linear.

This is used as the underlying implementation of EConstr.kind, which incurs a small constant overhead but changes the complexity class of all EConstr-using functions. In practice it does not matter except in degenerate examples.

### CODE REVIEW

**Comment by ppedrot** [comment]
@coqbot bench

**Comment by ppedrot** [comment]
New bench with whd_evar using the new structure. @coqbot bench

**Comment by ppedrot** [comment]
@coqbot bench

**Comment by ppedrot** [comment]
@coqbot ci minimize

**Comment by ppedrot** [comment]
Reduced version of the HoTT bug:
```coq
(* -*- mode: coq; coq-prog-args: ("-emacs" "-q" "-noinit" "-indices-matter" "-w" "-deprecated-native-compiler-option" "-native-compiler" "no" "-R" "/github/workspace/builds/coq/coq-failing/_build_ci/hott/theories" "HoTT" "-Q" "/github/workspace/cwd" "Top" "-Q" "/github/workspace/builds/coq/coq-failing/_build_ci/hott/contrib" "HoTT.Contrib" "-Q" "/github/workspace/builds/coq/coq-failing/_build_ci/hott/test" "HoTT.Tests" "-Q" "/github/workspace/builds/coq/coq-failing/_install_ci/lib/coq/user-contrib/Ltac2" "Ltac2" "-top" "HoTT.Modalities.ReflectiveSubuniverse") -*- *)

Declare ML Module "ltac_plugin".
Set Default Proof Mode "Classic".
Set Universe Polymorphism.
Unset Strict Universe Declaration.

Notation "A -> B" := (forall (_ : A), B) (at level 99, right associativity, B at level 200).

Notation compose := (fun g f x => g (f x)).

Inductive paths {A : Type} (a : A) : A -> Type := idpath : paths a a.

Definition pointwise_paths {A} {P : A -> Type} (f g : forall x, P x)
  := forall x, paths (f x) (g x).

Axiom In_internal@{i} : Type@{i} -> Type@{i}.
Axiom O_reflector@{i} : forall T : Type@{i}, Type@{i}.
Axiom to@{i} : forall (T : Type@{i}), T -> O_reflector@{i} T.
Axiom O_indpaths : forall {P Q : Type} (g h : O_reflector P -> Q)
             (p : pointwise_paths (compose g (to P)) (compose h (to P))), pointwise_paths g h.
Axiom O_indpaths_beta : forall {P Q : Type} (g h : O_reflector P -> Q)
             (p : pointwise_paths (compose g (to P)) (compose h (to P))) (x : P), paths (O_indpaths g h p (to P x)) (p x).
Axiom inO_to_O_retract : forall (T:Type) (mu : O_reflector T -> T),
  pointwise_paths (compose mu (@to T)) (fun x => x) -> In_internal T.

Definition inO_paths@{i} (S : Type@{i}) {S_inO : In_internal S} (x y : S) : In_internal (paths x y).
Proof.
simple refine (inO_to_O_retract@{i} _ _ _); intro u.
- assert (p : pointwise_paths (fun _ : O_reflector (paths x y) => x) (fun _=> y)).
  { refine (O_indpaths _ _ _); simpl.
    intro v; exact v. }
  exact (p u).
- simpl.
```

**Comment by ppedrot** [comment]
@coqbot bench

**Comment by ppedrot** [comment]
There is a slight constant runtime overhead to this PR, which can be observed in the extreme cases of the coq-performance-tests files. Nonetheless, it changes the complexity class of the EConstr.kind function, so I think that despite some minor regressions it should be merged, as justified by the oddorder speedup.

**Comment by ppedrot** [comment]
Note also that #19146 should totally remove the source of slowness of the extreme performance test exacerbated by this PR.

**Comment by ppedrot** [comment]
I deem this ready for merging. We can do better by exposing the new API and using it everywhere we fold / iter over econstrs to reduce some of the overhead, but this is minor improvements compared to the one brought by this PR.

File: engine/evd.ml [hunk 1]
```diff
@@ -1695,25 +1837,23 @@ module MiniEConstr = struct
...
-      begin match existential_opt_value sigma ev with
-        | None -> c
-        | Some c -> whd_evar sigma (mkCast (c, k, t))
-      end
+      let (h, knd) = Expand.kind sigma Expand.empty_handle c0 in
+      if Constr.kind c0 == knd then c
+      else mkCast (whd_kind sigma h knd, k, t)
     | _ -> c
+  and whd_kind sigma h knd =
+    (* we need to force the head as Kind.expand does not expand evar subterms *)
```
Comment (Janno) [inline engine/evd.ml:1855]:
```suggestion
    (* we need to force the head as Expand.kind does not expand evar subterms *)
```

**Review by Janno — COMMENTED** [review]
As somebody who has a lot of OCaml code containing fold/iter over econstrs I would prefer having the new api exposed directly.

**Comment by ppedrot** [comment]
@Janno if so I also need to write the corresponding iterators, that's quite a non-trivial amount of code to expose.

**Comment by Janno** [comment]
Oh. I thought it could be enough to expose the `Expand` module and have plugin developers figure out their own iterators.

**Comment by ppedrot** [comment]
The expand module is enough if you're willing to basically copy-paste the Constr equivalents, but it's a lot of code duplication.

**Comment by Janno** [comment]
I see. I withdraw my request then. Independent of further possible speedups this should definitely go in quickly.

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**